src/dialogs/copyPasteOwnership.tsx

```diff
@@ -0,0 +1,185 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+import React, { useEffect, useState } from 'react';
+
+import { AlertVariant } from "@patternfly/react-core/dist/esm/components/Alert";
+import { Button } from "@patternfly/react-core/dist/esm/components/Button";
+import { Form, FormGroup } from "@patternfly/react-core/dist/esm/components/Form";
+import { FormSelect, FormSelectOption } from "@patternfly/react-core/dist/esm/components/FormSelect";
+import { Modal, ModalVariant } from "@patternfly/react-core/dist/esm/components/Modal";
+import { Text, TextContent } from "@patternfly/react-core/dist/esm/components/Text";
+
+import cockpit from 'cockpit';
+import { FileInfo } from 'cockpit/fsinfo.ts';
+import type { Dialogs, DialogResult } from 'dialogs';
+import { superuser } from 'superuser';
+
+import { useFilesContext } from '../app.tsx';
+import type { ClipboardInfo } from '../app.tsx';
+import { get_owner_candidates } from '../ownership.tsx';
+
+const _ = cockpit.gettext;
+
+async function pasteAsOwner(clipboard: ClipboardInfo, dstPath: string, owner: string,
+    addAlert: (title: string, variant: AlertVariant, key: string, detail?: string) => void) {
+    try {
+        await cockpit.spawn([
+            "cp",
+            "--recursive",
+            ...clipboard.files.map(file => clipboard.path + "/" + file.name),
+            dstPath
+        ], { superuser: "try" });
+
+        await cockpit.spawn([
+            "chown",
+            "--recursive",
+            owner,
+            ...clipboard.files.map(file => dstPath + "/" + file.name),
+        ], { superuser: "try" });
+    } catch (err) {
+        const e = err as cockpit.BasicError;
+        addAlert(e.message, AlertVariant.danger, `${new Date().getTime()}`);
+
+        // cleanup potentially copied files in case of "chown" fail
+        try {
+            await cockpit.spawn([
+                "rm",
+                "-rf",
+                ...clipboard.files.map(file => dstPath + "/" + file.name)
+            ], { superuser: "try" });
+        } catch (ex) {
+            console.warn(`Failed to clean up copied files in ${dstPath}`, ex);
+        }
+    }
+}
+
+function makeCandidatesMap(currentUser: cockpit.UserInfo, cwdInfo: FileInfo, clipboard: ClipboardInfo) {
+    const map: Map<string, string> = new Map();
+    const candidates = get_owner_candidates(currentUser, cwdInfo);
+
+    // also add current ownership if it is same for all files (shallow check)
+    const firstFile = clipboard.files[0];
+    const firstOwnerStr = `${firstFile.user}:${firstFile.group}`;
+    if (clipboard.files.every(file => file.user === firstFile.user && file.group === firstFile.group)) {
+        candidates.add(firstOwnerStr);
+    }
```

Comment:
Hmm -- why the restriction to "same for all files"? It's harder to present in the UI (we can't give a definitive owner), but keeping the restrictions is especially useful if there are different owners -- i.e. you want to copy / move an entire hierarchy someplace else.

It shouldn't be too hard to add a `ClipboardInfo.mixed_owners: boolean` and showing that in the UI as "first owner (and others)" or so. Fine as follow-up material of course.